nibabel/streamlines/tractogram.py

```diff
@@ -101,15 +102,28 @@ def __init__(self, n_rows=0, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
     def __setitem__(self, key, value):
-        value = np.asarray(list(value))
+        dtype = np.float64
+
+        if isinstance(value, types.GeneratorType):
+            value = list(value)
+
+        if isinstance(value, np.ndarray):
+            dtype = value.dtype
+        elif not all(len(v) == len(value[0]) for v in value[1:]):
+            dtype = object
+
+        value = np.asarray(value, dtype=dtype)
 
         if value.ndim == 1 and value.dtype != object:
             # Reshape without copy
             value.shape = (len(value), 1)
 
-        if value.ndim != 2:
+        if value.ndim != 2 and value.dtype != object:
             raise ValueError('data_per_streamline must be a 2D array.')
 
+        if value.dtype == object and not all(isinstance(v, Iterable) for v in value):
+            raise ValueError('data_per_streamline must be a 2D array')
```

Comment:
Is it possible to test this case? I think passing something like `np.array([1, np.array([0, 1])], dtype=object)`?